polka-index/RFC.md

```diff
@@ -0,0 +1,131 @@
+- **pubsub**: A protocol where peers congregate around topics they are interested in.
+- **Dialing**: The process in [libp2p](https://docs.libp2p.io/) where one peer establishes a direct connection to another peer using its Peer ID.
+
+## 3. System Overview
+
+### 3.1 Components
+
+#### Storage Providers Component
+
+- Connect to the P2P network using unique Peer IDs.
+- Broadcast their Peer IDs using pubsub so they can be discovered by the Polka-Index.
+- Maintain a Deal Database containing the CIDs and metadata for the files they store.
+- Respond to Polka-Index queries via dialing.
+
+#### Polka-Index Component
+
+- Connects to the P2P network using its own unique Peer ID.
+- Detects storage providers through pubsub broadcasts.
+- Dials storage providers directly to query for CID mappings.
+- Maintains a CID-to-Peer ID mapping database.
+- Provides an HTTP API for users to request CID-to-Peer ID mappings.
+
+#### P2P Network
+
+- A communication layer powered by [libp2p](https://docs.libp2p.io/) that supports Peer ID discovery (via pubsub) and direct connections (via dialing).
+Users
+- Connect to the Polka-Index via HTTP to find out which Peer ID is associated with a specific CID.
+
+### 3.2 How It Works
+
+1. Storage providers connect to the P2P network and broadcast their Peer IDs using pubsub.
+2. Polka-Index listens for these broadcasts and keeps track of active storage providers.
+3. Users send an HTTP request to the Polka-Index, asking for the Peer ID associated with a specific CID.
+4. If the CID isn’t already in its database, the Polka-Index dials storage providers directly to ask for the mapping.
+5. The storage provider responds with the CID-to-Peer ID mapping.
+6. Polka-Index stores the mapping for future use and returns the Peer ID to the user.
+
+### 3.3 Architecture Diagram
+
+The following diagram illustrates the architecture of Polka-index:
+
+![architecture](assets/Polka-indexer.svg)
+
+### 4. Architectural Details
+
+#### Storage Providers Architecture
+
+- **Peer ID**: Each storage provider generates a Peer ID when it connects to the P2P network. This ID is broadcast via pubsub for discovery.
+- **Deal Database**: A lightweight database where storage providers keep records of CIDs and file metadata. File storage itself is external to the P2P network.
+- **Dialing**: Responds to direct queries from the Polka-Index to provide CID-to-Peer ID mappings.
+
+#### Polka-Index Architecture
+
+- **Peer ID**: The Polka-Index has its own Peer ID for identifying itself in the P2P network.
+- **pubsub Discovery**: It listens for Peer ID broadcasts to discover active storage providers.
+- **CID Mapping Database**: Stores CID-to-Peer ID mappings retrieved from storage providers. This database powers the HTTP API for user queries.
+- **HTTP API**: Provides a simple interface for users:
```

Comment:
In this case, it might be better to stream the "fetch candidates". The indexer might also want to find out if there are any new providers storing the data. You can't be sure that you have the best ones cached.